### PR TITLE
Add layout content size controls to global styles

### LINKED ISSUE #35955: Allow users to edit the "default layout" in the Global Styles panel
Whenever you see a block with a "Layout" panel, you're able to select "Inherit default layout." This setting pulls from the theme options to provide a default layout. 

It seems natural that users would be able to edit these values in the "Layout" section of Global Styles, but they're not there currently:

Screen Shot 2021-10-26 at 8 20 37 AM

(I searched before posting this, and was surprised that there wasn't an existing issue for this?)

### PR DESCRIPTION
**Pull request by tellthemachines**
## What?

Adds content size and wide size controls to global styles, under "Layout"

## Why?

Fixes #35955.

## How?

* Enables `contentSize` and `wideSize` in `ROOT_BLOCK_SUPPORTS`;
* Adds controls to Dimensions panel in the root Layout section of Global Styles sidebar.

## Testing Instructions

1. In the global styles panel, go to Layout and check that Content and Wide size inputs are visible under "Dimensions".
2. Add values to those fields and check that content sizes update sitewide.
3. Check that values can be reset from the kebab menu in the Dimensions panel.

## Screenshots or screencast 

Screen Shot 2022-07-15 at 5 45 23 pm

Screen Shot 2022-07-15 at 5 45 16 pm

### CODE REVIEW

**Review by andrewserong — COMMENTED** [review]
Oh, this is testing quite nicely so far! I was wondering at first if the `HStack` component might help simplify things a bit versus the `Flex` component (it doesn't seem to need the children the be `FlexItem`s?), but I don't think it helps directly with the icon positioning. Shame that UnitControl's positioning doesn't appear to neatly allow additional elements at the same vertical position as the input field 🤔

The UI here also demonstrates nicely a point you made in the root CSS PR that these controls seem to fit logically with padding and other dimensions controls, they look good alongside each other, all under a Layout panel. Not sure what that means for future iterations of the block level inspector controls, but at least from my perspective, these controls look nicely at home in this panel! 👍

Not really related to this PR, but I noticed while testing that the Reset to Defaults button in the dropdown correctly resets the values, but not the ToolsPanel state — so the ellipsis menu thinks that the value is still customised. It doesn't appear to actually cause any issues so probably not something to worry about in this PR, but just thought I'd mention it:

**Comment by tellthemachines** [comment]
Thanks for testing @andrewserong ! I hadn't thought of `HStack`; I tried it but as you mentioned the icon positioning isn't quite right. It's working well with Flex now that I added alignment, but I suspect that's mostly because of the `div` wrapper that `FlexItem` adds around it 😅 

As a side note, we can use the `as` prop on `FlexItem` to render children without the `div`, but for some reason doing that with `Icon` strips out the contents of the `svg` element. Seems like a bug, will test further.

Another option would be to add styles directly instead of using these components, and that would allow us to make micro-adjustments to margins as needed.

> Not really related to this PR, but I noticed while testing that the Reset to Defaults button in the dropdown correctly resets the values, but not the ToolsPanel state — so the ellipsis menu thinks that the value is still customised. It doesn't appear to actually cause any issues so probably not something to worry about in this PR, but just thought I'd mention it:

Ah, well spotted! Testing on trunk, it seems like the global reset doesn't change the padding value in the input either - it does reset it, but the number only disappears after save and refresh. Might be best to look into that behaviour separately.

**Comment by tellthemachines** [comment]
Update: opened #42380 for FlexItem swallowing its children.

File: packages/edit-site/src/components/global-styles/dimensions-panel.js [hunk 1]
```diff
@@ -141,6 +178,66 @@ export default function DimensionsPanel( { name } ) {
 
 	return (
 		<ToolsPanel label={ __( 'Dimensions' ) } resetAll={ resetAll }>
+			<Flex>
```
Comment (andrewserong) [inline packages/edit-site/src/components/global-styles/dimensions-panel.js]: It looks like this row might need to be behind a check for `showContentSizeControl || showSideSizeControl`, otherwise in the Layout panel at the block level, there is a larger space due to the gap surrounding the empty `Flex` component:

image

File: packages/edit-site/src/components/global-styles/dimensions-panel.js [hunk 2]
```diff
@@ -99,6 +120,22 @@ export default function DimensionsPanel( { name } ) {
 		],
 	} );
 
+	const [ contentSizeValue, setContentSizeValue ] = useSetting(
+		'layout.contentSize',
+		name
+	);
+
+	const hasContentSizeValue = () => !! contentSizeValue;
```
Comment (andrewserong) [inline packages/edit-site/src/components/global-styles/dimensions-panel.js]: Taking another look at this check, I'm wondering if part of the issue with the ToolsPanel menu state is that if the `contentSizeValue` matches the value after reset, then this will return `true`. So if my content width is set to `840px` and I change it to `900px` and then reset it, the value here is `840px`, and so the menu item thinks there's still a custom value?

That said, there seems to be something else funny going on, because when I first load the site editor, the ToolsPanel seems to be showing padding as customised, too 🤔

image

The weird behaviour does appear to be on `trunk`, too, so not sure if it's worth attempting to fix now? I think on `trunk` it's a little less visible because it seems to be working fine at the block level, and we currently only have the `Padding` dimensions control made visible at the root level 🤔

**Review by andrewserong — COMMENTED** [review]
The positioning in the UI's looking good now @tellthemachines! Do we need to add the explanatory text from the editor UI for the content/wide sizes, too?

Updating the content and wide size at the root level is working well, but the reset behaviour is still a bit unreliable for me. Also, I found an edge case with blocks that don't display content/wide size, and added a comment about the weird ToolsPanel menu behaviour — unfortunately I couldn't quite work out what's going on there 🤔

Another thought is that the `resetAll` function passed to the `ToolsPanel` should probably include the reset calls for the content and wide sizes, too... but probably only if the `showContentSizeControl` and `showWideSizeControl` variables are truthy?

Happy to do more testing!

File: packages/edit-site/src/components/global-styles/dimensions-panel.js (hunk 1 above)
Comment (tellthemachines) [reply packages/edit-site/src/components/global-styles/dimensions-panel.js]: Oh, yes, definitely!

File: packages/edit-site/src/components/global-styles/dimensions-panel.js [hunk 3]
```diff
@@ -141,6 +178,71 @@ export default function DimensionsPanel( { name } ) {
...
+			{ ( showContentSizeControl || showWideSizeControl ) && (
+				<Flex>
+					{ showContentSizeControl && (
+						<FlexItem>
+							<ToolsPanelItem
+								label={ __( 'Content size' ) }
+								hasValue={ hasContentSizeValue }
+								onDeselect={ resetContentSizeValue }
+								isShownByDefault={ true }
+							>
```
Comment (aaronrobertshaw) [inline packages/edit-site/src/components/global-styles/dimensions-panel.js]: `ToolsPanelItem` is polymorphic as well so you should be able to just add the `as={ FlexItem }` to it right?

File: packages/edit-site/src/components/global-styles/dimensions-panel.js (hunk 2 above)
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/global-styles/dimensions-panel.js]: > So if my content width is set to 840px and I change it to 900px and then reset it, the value here is 840px, and so the menu item thinks there's still a custom value?

It appears that the menu is reflecting exactly what the `hasContentSizeValue` callback is telling it. Any truthy value for `contentSizeValue` is going to state that the `ToolsPanelItem` has a customized value.

If I understand the situation correctly, there's always a root value so the logic here needs to be more nuanced. It should be checking if the user has set a value i.e. check the style value from the `user` origin rather than the merged styles aka `all` origin.

The [border panel checks the `user` origin border styles](https://github.com/WordPress/gutenberg/blob/8b1492131355aa6d42272faba25b841f0cd4509d/packages/edit-site/src/components/global-styles/border-panel.js#L100) to reflect the whether the user has customized any values e.g. `const [ userBorderStyles ] = useStyle( 'border', name, 'user' );`

It would be a worthwhile follow up to give these global styles panels an audit to check they all behave appropriately. I know things have evolved rapidly here on multiple fronts so its entirely possible a few gremlins crept in.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for your work on this @tellthemachines 👍 

In terms of the content/width size, it is testing pretty well for me. I hit similar issues as @andrewserong however with the new controls/panel behaviour.

Before digging too deep into things, it might be worth highlighting here a recent [issue](https://github.com/WordPress/gutenberg/issues/42455) found that impacts resetting input control based components for which @stokesman already has a fix up for in https://github.com/WordPress/gutenberg/pull/42484. 

Cherry picking the fix from that PR fixed a couple of resetting issues encountered while testing this one.

> Not really related to this PR, but I noticed while testing that the Reset to Defaults button in the dropdown correctly resets the values, but not the ToolsPanel state

Global Styles' option to "reset to defaults" clears all the values independently. Given it happens outside the `ToolsPanel` callbacks there is nothing to trigger its internal state update. (Would it be too hacky to force a new `ToolsPanel` when the global styles are reset?)

In some past discussions it was raised that clearing a value from the field could be a customization and therefore we couldn't really catch the fact the field was empty and automatically remove the customized status reflected in the panel menu. The thinking at the time was that this wasn't a real issue. I'm not sure if that needs revisiting though via a follow-up.

Long story, short. I don't think the "reset to defaults" issue should block this PR at all.

I've also left a couple of other inline comments that might help smooth out the customized value display or reduce some markup. Hope that helps.

File: packages/edit-site/src/components/global-styles/dimensions-panel.js (hunk 2 above)
Comment (tellthemachines) [reply packages/edit-site/src/components/global-styles/dimensions-panel.js]: Checking for user styles makes sense! Updated.

File: packages/edit-site/src/components/global-styles/dimensions-panel.js (hunk 3 above)
Comment (tellthemachines) [reply packages/edit-site/src/components/global-styles/dimensions-panel.js]: Ahhhh thanks hadn't thought of that!

**Comment by tellthemachines** [comment]
Thanks for the reviews, folks!

> Do we need to add the explanatory text from the editor UI for the content/wide sizes, too?

I think the text we see currently in the block Layout panel ("Customize the width for all elements that are assigned to the center or wide columns.") is a bit confusing even in the block context, and more so here because we're setting widths site-wide here, so "customize" in this case just means "set a new global value". Is it just me? Perhaps we should rethink that copy, and think of a specific one for this global control. Would something like "Set the width of the main content area" work here?

> Also, I found an edge case with blocks that don't display content/wide size

Was that the Flex markup showing, or did I miss something else?

> Before digging too deep into things, it might be worth highlighting here a recent https://github.com/WordPress/gutenberg/issues/42455 found that impacts resetting input control based components for which @stokesman already has a fix up for in https://github.com/WordPress/gutenberg/pull/42484.

Ah, good to know! If that one gets merged soon I'll rebase this PR.

> Would it be too hacky to force a new ToolsPanel when the global styles are reset?

I'm not sure. Thinking about the UX, it's unlikely someone will go into the individual reset panels after resetting all, unless the global reset didn't work? Maybe we can look at that as a separate task and consider the options.

One thing I'm not sure about: currently, these controls only display if the theme has contentSize/wideSize set in its `theme.json`. Is this the desired state of things? Or do we want it to be possible to set content widths even if the theme doesn't support them?

**Comment by aaronrobertshaw** [comment]
> One thing I'm not sure about: currently, these controls only display if the theme has contentSize/wideSize set in its theme.json. Is this the desired state of things? Or do we want it to be possible to set content widths even if the theme doesn't support them?

My understanding is that one of the features of theme.json is that it allows theme authors to lock down what is available in the Global Styles UI. For example, they could lock down editing color palettes, prevent users from applying custom borders etc.

If a theme has not opted to set `contentSize`/`wideSize` is that a choice not to expose them to users? Alternatively, should we have explicit settings properties in the theme.json to flag whether custom `contentSize` or `wideSize` values are accepted and the UI shown? I'm thinking along the lines of `customFontSize`.

**Comment by andrewserong** [comment]
>  Is it just me?

Not just you! I find it confusing, too. If it's too awkward to do here, perhaps tweaking how and when we display explanatory information for this feature is best explored in a follow-up / separate PR at some stage? But I quite like "Set the width of the main content area" personally 👍

> Was that the Flex markup showing, or did I miss something else?

Nope, that was all!

> Alternatively, should we have explicit settings properties in the theme.json to flag whether custom contentSize or wideSize values are accepted and the UI shown?

Oh, that's an interest idea. I don't think I'd even really thought of locking down this part of the UI (I'd mostly been thinking about when themes want to lock down the post editor, rather than global styles). Perhaps a boolean in theme settings that defaults to `true` (that it's always editable) and then the theme can set it to `false` if need be? Personally I like the idea of folks being able to edit content / wide widths in any theme (along the lines of "designing in the site editor"), but yeah, it might be good to consider theme builder's requirements here 🤔

**Comment by tellthemachines** [comment]
> My understanding is that one of the features of theme.json is that it allows theme authors to lock down what is available in the Global Styles UI.

Hmmm, in that case it would make sense for themes be able to both set content sizes _and_ stop users from overriding them. If so, and if

> Alternatively, should we have explicit settings properties in the theme.json to flag whether custom contentSize or wideSize values are accepted and the UI shown?

this should probably also apply to the individual block controls, right?

And if we had something like this, we needn't assume that a theme without content sizes doesn't want users to be able to set them at all.

**Comment by tellthemachines** [comment]
Update on the description: I tried to add a paragraph inside the `ToolsPanel` as indicated in the [component readme example](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/tools-panel/tools-panel/README.md#usage) but it doesn't work with the styling:

Screen Shot 2022-07-20 at 11 54 31 am

I've created #42543 to report what seems to be a `ToolsPanel` bug.

I guess we can add the description as a separate PR once the issue is fixed?

I'm thinking the bit about allowing themes to explicitly control whether content and wide sizes are editable can also be tackled separately, as it'll probably generate a bit of discussion 😅 

If that's the case, I think all other feedback on this PR has been addressed!

**Review by andrewserong — APPROVED** [review]
> If that's the case, I think all other feedback on this PR has been addressed!

Yes, this PR is working nicely for me now, and I agree I think adding messaging and figuring out the appropriate behaviour for showing/hiding the controls are good candidates for follow-up PRs.

✅ Adjusting content and wide width is now treated in the ToolsPanel menu as a changed value if a user set value exists
✅ Clearing each value in the menu individually works correctly
✅ Clearing via the _Reset all_ option works correctly for these two values
✅ Styling and behaviour of the Dimensions panel at the block level in global styles is as on `trunk`

LGTM, thanks for all the follow-up tweaks! 🎉

**Comment by aaronrobertshaw** [comment]
I think we might be overcomplicating the addition of these controls to the panel. 🤔 

Am I correct in understanding that the wrapping `Flex` was only to make the two controls share the same row in the panel? 

The `ToolsPanel` has a two-column grid primarily for this purpose. `ToolsPanelItem` components default to spanning both columns given that was the majority use-case at the time the design was being implemented.

The same way the `FlexItem` currently provides a wrapper to assist in the icon positioning, a simple wrapper could be added within a `HStack` to achieve the same result. Not much of a difference but does mean slightly less fluff in the markup.

Finally, for the help text, this is a bit of an edge case. Normally, help text would relate to a single control and be included as a child within the `ToolsPanelItem`. Within the block editor inspector controls we don't have a lot of panels beginning with a description like some of the Global Styles panels so there hasn't been anything baked into the `ToolsPanel` for that.

In the example diff below I added a simple `span-columns` class for the global styles that could be shared across any top-level non-`ToolsPanelItem` components.


<details>
<summary>Click for example diff</summary>

```diff
diff --git a/packages/edit-site/src/components/global-styles/dimensions-panel.js b/packages/edit-site/src/components/global-styles/dimensions-panel.js
index a2f8822d66..e7af30623c 100644
--- a/packages/edit-site/src/components/global-styles/dimensions-panel.js
+++ b/packages/edit-site/src/components/global-styles/dimensions-panel.js
@@ -6,10 +6,10 @@ import {
 	__experimentalToolsPanel as ToolsPanel,
 	__experimentalToolsPanelItem as ToolsPanelItem,
 	__experimentalBoxControl as BoxControl,
+	__experimentalHStack as HStack,
 	__experimentalUnitControl as UnitControl,
 	__experimentalUseCustomUnits as useCustomUnits,
-	Flex,
-	FlexItem,
+	__experimentalView as View,
 } from '@wordpress/components';
 import { __experimentalUseCustomSides as useCustomSides } from '@wordpress/block-editor';
 import { Icon, positionCenter, stretchWide } from '@wordpress/icons';
@@ -193,65 +193,59 @@ export default function DimensionsPanel( { name } ) {
 	return (
 		<ToolsPanel label={ __( 'Dimensions' ) } resetAll={ resetAll }>
 			{ ( showContentSizeControl || showWideSizeControl ) && (
-				<Flex>
-					{ showContentSizeControl && (
-						<ToolsPanelItem
-							as={ FlexItem }
-							label={ __( 'Content size' ) }
-							hasValue={ hasUserSetContentSizeValue }
-							onDeselect={ resetContentSizeValue }
-							isShownByDefault={ true }
-						>
-							<Flex align="flex-end">
-								<FlexItem>
-									<UnitControl
-										label={ __( 'Content' ) }
-										labelPosition="top"
-										__unstableInputWidth="80px"
-										value={ contentSizeValue || '' }
-										onChange={ ( nextContentSize ) => {
-											setContentSizeValue(
-												nextContentSize
-											);
-										} }
-										units={ units }
-									/>
-								</FlexItem>
-								<FlexItem>
-									<Icon icon={ positionCenter } />
-								</FlexItem>
-							</Flex>
-						</ToolsPanelItem>
-					) }
-
-					{ showWideSizeControl && (
-						<ToolsPanelItem
-							as={ FlexItem }
-							label={ __( 'Wide size' ) }
-							hasValue={ hasUserSetWideSizeValue }
-							onDeselect={ resetWideSizeValue }
-							isShownByDefault={ true }
-						>
-							<Flex align="flex-end">
-								<FlexItem>
-									<UnitControl
-										label={ __( 'Wide' ) }
-										labelPosition="top"
-										__unstableInputWidth="80px"
-										value={ wideSizeValue || '' }
-										onChange={ ( nextWideSize ) => {
-											setWideSizeValue( nextWideSize );
-										} }
-										units={ units }
-									/>
-								</FlexItem>
-								<FlexItem>
-									<Icon icon={ stretchWide } />
-								</FlexItem>
-							</Flex>
-						</ToolsPanelItem>
-					) }
-				</Flex>
+				<span className="span-columns">
+					Set the width of the main content area.
+				</span>
+			) }
+			{ showContentSizeControl && (
+				<ToolsPanelItem
+					className="single-column"
+					label={ __( 'Content size' ) }
+					hasValue={ hasUserSetContentSizeValue }
+					onDeselect={ resetContentSizeValue }
+					isShownByDefault={ true }
+				>
+					<HStack alignment="flex-end">
+						<UnitControl
+							label={ __( 'Content' ) }
+							labelPosition="top"
+							__unstableInputWidth="80px"
+							value={ contentSizeValue || '' }
+							onChange={ ( nextContentSize ) => {
+								setContentSizeValue( nextContentSize );
+							} }
+							units={ units }
+						/>
+						<View>
+							<Icon icon={ positionCenter } />
+						</View>
+					</HStack>
+				</ToolsPanelItem>
+			) }
+			{ showWideSizeControl && (
+				<ToolsPanelItem
+					className="single-column"
+					label={ __( 'Wide size' ) }
+					hasValue={ hasUserSetWideSizeValue }
+					onDeselect={ resetWideSizeValue }
+					isShownByDefault={ true }
+				>
+					<HStack alignment="flex-end">
+						<UnitControl
+							label={ __( 'Wide' ) }
+							labelPosition="top"
+							__unstableInputWidth="80px"
+							value={ wideSizeValue || '' }
+							onChange={ ( nextWideSize ) => {
+								setWideSizeValue( nextWideSize );
+							} }
+							units={ units }
+						/>
+						<View>
+							<Icon icon={ stretchWide } />
+						</View>
+					</HStack>
+				</ToolsPanelItem>
 			) }
 			{ showPaddingControl && (
 				<ToolsPanelItem
diff --git a/packages/edit-site/src/components/sidebar/style.scss b/packages/edit-site/src/components/sidebar/style.scss
index b3992b7774..41960d3ee0 100644
--- a/packages/edit-site/src/components/sidebar/style.scss
+++ b/packages/edit-site/src/components/sidebar/style.scss
@@ -60,6 +60,10 @@
 	grid-column: span 1;
 }
 
+.edit-site-global-styles-sidebar .components-tools-panel .span-columns {
+	grid-column: 1 / -1;
+}
+
 .edit-site-global-styles-sidebar__blocks-group {
 	padding-top: $grid-unit-30;
 	border-top: $border-width solid $gray-200;

```
</details>

After applying the above diff, this is what I get:

Screen Shot 2022-07-20 at 3 07 51 pm

One thing to note is that the panel's CSS grid means spacing between the "Set the width of the main content area." text and the related controls. That might not be desired so please take the suggested diff with a grain of salt.

If nothing else, I hope that helps illustrate a different approach to adding controls via the `ToolsPanel`.

**Comment by tellthemachines** [comment]
> In the example diff below I added a simple span-columns class for the global styles that could be shared across any top-level non-ToolsPanelItem components.

Oooh I see that works well! Would be good to add some tips along those lines to the `ToolsPanel` readme. Also please feel free to push that diff to this branch 😄

**Comment by aaronrobertshaw** [comment]
> Also please feel free to push that diff to this branch

Done 👍 

> Would be good to add some tips along those lines to the ToolsPanel readme.

I'll make that happen 🙂

File: packages/edit-site/src/components/global-styles/dimensions-panel.js [hunk 4]
```diff
@@ -137,10 +186,67 @@ export default function DimensionsPanel( { name } ) {
...
 		resetGapValue();
+		resetContentSizeValue();
+		resetWideSizeValue();
 	};
 
 	return (
 		<ToolsPanel label={ __( 'Dimensions' ) } resetAll={ resetAll }>
+			{ ( showContentSizeControl || showWideSizeControl ) && (
+				<span className="span-columns">
+					Set the width of the main content area.
```
Comment (andrewserong) [inline packages/edit-site/src/components/global-styles/dimensions-panel.js]: Apologies if you're already in the middle of tweaking this, but just a quick drive-by, looks like this needs a `__()` call 🙂

Comment (aaronrobertshaw) [reply packages/edit-site/src/components/global-styles/dimensions-panel.js]: Thanks. Fixed.

The dangers of pushing quick diffs from testing. ☠️ 🙂

**Comment by paaljoachim** [comment]
Dropping by for a quick look.
Viewing the initial gif I see that reset option has been hidden away behind the 3 dots menu.
This is though very hidden away. Having the option as it was earlier fully visible made it clear that it is possible to reset the values that have been added. Having it inside the 3 dot menu it is easy to think that the option has been removed.

Some options are natural to add inside the 3 dot menu. Other options should in general always be visible.
